File: Fitness Tracker/StartWorkout.m
//
//  StartWorkout.m
//  Fitness Tracker
//
//  Created by Maruf Nebil Ogunc on 8.08.2018.
//  Copyright © 2018 Maruf Nebil Ogunc. All rights reserved.
//

#import "StartWorkout.h"
#import "SeparatedExerciseCell.h"

@implementation StartWorkout
@synthesize workoutNo,exercisesTableView;


- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self.view setBackgroundColor:[UIColor whiteColor]];
    
    //Getting size of the device
    CGRect screenRect = [[UIScreen mainScreen] bounds];

    //getting workouts array from user defaults
    NSMutableArray * workoutsArray = [[[NSUserDefaults standardUserDefaults] arrayForKey:@"savedWorkouts"] mutableCopy];

    //removing .json extension
    NSString * workoutNameWithoutExtension = [[workoutsArray objectAtIndex:self.workoutNo] stringByDeletingPathExtension];

    //Setting navigation bar title
    [self.navigationController.navigationBar setTitleTextAttributes: @{NSForegroundColorAttributeName:[UIColor blackColor], NSFontAttributeName:[UIFont fontWithName:@"Metropolis-Bold" size:20]}];
    self.navigationItem.title = workoutNameWithoutExtension;

    //Creating cancel button which will close the modal
    UIBarButtonItem *cancelButton = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCancel target:self action:@selector(cancelButtonHit)];
    self.navigationItem.rightBarButtonItem = cancelButton;

    //Creating table view which will show exercises
    exercisesTableView = [[UITableView alloc] initWithFrame:CGRectMake(0, 0, screenRect.size.width, screenRect.size.height - 80)];
    [exercisesTableView setBackgroundColor:[UIColor whiteColor]];
    exercisesTableView.delegate = self;
    exercisesTableView.dataSource = self;
    [self.view addSubview:exercisesTableView];

    //This button will start and pause the workout
    startPauseWorkoutButton = [UIButton buttonWithType:UIButtonTypeCustom];
    [startPauseWorkoutButton setTitle:@"START" forState:UIControlStateNormal];
    [startPauseWorkoutButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    [startPauseWorkoutButton.titleLabel setFont:[UIFont fontWithName:@"Metropolis-Medium" size:18.0]];
    startPauseWorkoutButton.titleLabel.textAlignment = NSTextAlignmentCenter;
    [startPauseWorkoutButton setBackgroundColor:[UIColor colorWithRed:0 green:179.0/255.0 blue:85.0/255.0 alpha:1]];
    startPauseWorkoutButton.frame = CGRectMake(0, screenRect.size.height - 80, screenRect.size.width, 80);
    [startPauseWorkoutButton addTarget:self action:@selector(startPauseWorkout) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:startPauseWorkoutButton];

    //workout is not active default
    isWorkoutActive = false;
    
    //initialize exercises array
    separatedExercisesArray = [[NSMutableArray alloc] init];

    [self readWorkoutJSONbyName:[workoutsArray objectAtIndex:self.workoutNo]];

}

-(void)startPauseWorkout{
    
    if (isWorkoutActive) {
        isWorkoutActive = false;
        [startPauseWorkoutButton setBackgroundColor:[UIColor colorWithRed:0 green:179.0/255.0 blue:85.0/255.0 alpha:1]];
        [startPauseWorkoutButton setTitle:@"START" forState:UIControlStateNormal];
        [countdownTimer invalidate];
        
        //changing color of the cell
        int indexOfObject = [self getIndexOfFirstUncompletedExercise];
        NSDictionary * separatedExerciseDict = [separatedExercisesArray objectAtIndex:indexOfObject];
        NSMutableDictionary * newDict = [[NSMutableDictionary alloc] init];
        [newDict addEntriesFromDictionary:separatedExerciseDict];
        [newDict setObject:@2 forKey:@"status"];
        [separatedExercisesArray replaceObjectAtIndex:indexOfObject withObject:newDict];
        [exercisesTableView reloadData];

    }else{
        isWorkoutActive = true;
        [startPauseWorkoutButton setBackgroundColor:[UIColor colorWithRed:229.0/255.0 green:93.0/255.0 blue:41.0/255.0 alpha:1]];
        [startPauseWorkoutButton setTitle:@"PAUSE" forState:UIControlStateNormal];
        countdownTimer = [NSTimer scheduledTimerWithTimeInterval:1/30 target:self selector:@selector(updateTimer) userInfo:nil repeats:YES];
    }
    
}

-(int)getIndexOfFirstUncompletedExercise{
    
    for (int i = 0; i < separatedExercisesArray.count; i++) {
        
        NSDictionary * separatedExerciseDict = [separatedExercisesArray objectAtIndex:i];
        int duration = [[separatedExerciseDict objectForKey:@"duration"] intValue];
        
        if(duration > 0){
            return i;
        }
    }

    return -1;

}

-(void)updateTimer{
    
    int indexOfObject = [self getIndexOfFirstUncompletedExercise];

    //Reducing duration of the exercise
    NSDictionary * separatedExerciseDict = [separatedExercisesArray objectAtIndex:indexOfObject];
    int duration = [[separatedExerciseDict objectForKey:@"duration"] intValue];
    duration--;
    
    if (duration > 0) {
        
        NSMutableDictionary * newDict = [[NSMutableDictionary alloc] init];
        [newDict addEntriesFromDictionary:separatedExerciseDict];
        [newDict setObject:[NSString stringWithFormat:@"%d",duration] forKey:@"duration"];
        [newDict setObject:@1 forKey:@"status"];
        
        [separatedExercisesArray replaceObjectAtIndex:indexOfObject withObject:newDict];
        [exercisesTableView reloadData];

    }else{//when exercise is finished, remove it from the array
        
        [separatedExercisesArray removeObjectAtIndex:indexOfObject];
        [exercisesTableView reloadData];

    }
    
}

-(void)readWorkoutJSONbyName:(NSString*)name{
    
    NSString* filePath = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) objectAtIndex:0];
    NSString* fileName = name;
    NSString* fileAtPath = [filePath stringByAppendingPathComponent:fileName];
    
    if ([[NSFileManager defaultManager] fileExistsAtPath:fileAtPath]) {
        
        //Fill the array with the exercises
        NSMutableArray* rawExercisesArray = [[NSMutableArray alloc] init];
        rawExercisesArray = [NSJSONSerialization JSONObjectWithData:[NSData dataWithContentsOfFile:fileAtPath] options:NSJSONReadingMutableContainers error:nil];
        [self seperateExercisesBySetsAndRests:rawExercisesArray];
    }else{
        NSLog(@"File don't exist");
    }
}

-(void)seperateExercisesBySetsAndRests:(NSMutableArray*)rawExercisesArray{
    //NSLog(@"RAW%@",rawExercisesArray);
    
    for (int i = 0; i < rawExercisesArray.count; i++) {
        NSDictionary * exerciseDict = [rawExercisesArray objectAtIndex:i];
        
        if([[exerciseDict objectForKey:@"isCardio"] boolValue]){
            
            NSDictionary * separatedExercise = [[NSDictionary alloc] init];
            int cardioSeconds = [[exerciseDict objectForKey:@"cardio_minutes"] intValue] * 60;
            separatedExercise = @{ @"name" : [exerciseDict objectForKey:@"name"], @"duration" : [NSString stringWithFormat:@"%d",cardioSeconds], @"status" : @0};
            [separatedExercisesArray addObject:separatedExercise];
            
        }else{
            int sets = [[exerciseDict objectForKey:@"sets"] intValue];
            
            for (int i = 0; i < sets; i++) {
                
                NSDictionary * separatedExercise = [[NSDictionary alloc] init];
                separatedExercise = @{ @"name" : [exerciseDict objectForKey:@"name"], @"duration" : [exerciseDict objectForKey:@"rest"], @"reps" : [exerciseDict objectForKey:@"reps"], @"status" : @0};
                [separatedExercisesArray addObject:separatedExercise];

                if (i != sets - 1) {//don't add rest time for last move
                
                    NSDictionary * separatedExercise = [[NSDictionary alloc] init];
                    separatedExercise = @{ @"name" : @"Rest", @"duration" : [exerciseDict objectForKey:@"rest"], @"status" : @0};
                    [separatedExercisesArray addObject:separatedExercise];

                }
            }
            
        }
        
        //adding 90 seconds for rest and preparation
        if (i != rawExercisesArray.count - 1) {//don't add rest time for last exercise
            
            NSDictionary * separatedExercise = [[NSDictionary alloc] init];
            separatedExercise = @{ @"name" : @"Rest and Preparation", @"duration" : @"90", @"status" : @0};
            [separatedExercisesArray addObject:separatedExercise];
        }
    }
    
    //NSLog(@"SEP%@",separatedExercisesArray);

}

#pragma mark - Table View Data source
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return separatedExercisesArray.count;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView{
    return 1;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 70;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *cellIdentifier = @"separatedCell";
    
    SeparatedExerciseCell *cell = (SeparatedExerciseCell*)[tableView dequeueReusableCellWithIdentifier:cellIdentifier];
    if (cell == nil) {
        cell = [[SeparatedExerciseCell alloc] initWithStyle: UITableViewCellStyleDefault reuseIdentifier:cellIdentifier];
    }
    
    cell.selectionStyle = UITableViewCellSelectionStyleNone;

    NSDictionary * separatedExerciseDict = [separatedExercisesArray objectAtIndex:indexPath.row];
    int duration = [[separatedExerciseDict objectForKey:@"duration"] intValue];
    int minutes = duration / 60;
    int seconds = duration % 60;
    cell.countdownLabel.text = [NSString stringWithFormat:@"%d:%02d", minutes, seconds];
    cell.exerciseNameLabel.text = [separatedExerciseDict objectForKey:@"name"];
    
    if([separatedExerciseDict objectForKey:@"reps"]){
        cell.repsLabel.hidden = NO;
        cell.repsStaticLabel.hidden = NO;
        cell.repsLabel.text = [separatedExerciseDict objectForKey:@"reps"];
    }else{
        cell.repsLabel.hidden = YES;
        cell.repsStaticLabel.hidden = YES;
    }
    
    if ([[separatedExerciseDict objectForKey:@"status"] intValue] == 0) {//default status
        cell.backgroundColor = [UIColor whiteColor];
        cell.countdownLabel.textColor = [UIColor blackColor];
        cell.exerciseNameLabel.textColor = [UIColor blackColor];
        cell.exerciseNameLabel.font = [UIFont fontWithName: @"Metropolis-Medium" size: 20.0f];
    }else if ([[separatedExerciseDict objectForKey:@"status"] intValue] == 1){//active
        cell.backgroundColor = [UIColor colorWithRed:0 green:179.0/255.0 blue:85.0/255.0 alpha:1];
        cell.countdownLabel.textColor = [UIColor whiteColor];
        cell.exerciseNameLabel.textColor = [UIColor whiteColor];
        cell.exerciseNameLabel.font = [UIFont fontWithName: @"Metropolis-Bold" size: 20.0f];
    }else if ([[separatedExerciseDict objectForKey:@"status"] intValue] == 2){//paused
        cell.backgroundColor = [UIColor colorWithRed:229.0/255.0 green:93.0/255.0 blue:41.0/255.0 alpha:1];
        cell.countdownLabel.textColor = [UIColor whiteColor];
        cell.exerciseNameLabel.textColor = [UIColor whiteColor];
        cell.exerciseNameLabel.font = [UIFont fontWithName: @"Metropolis-Bold" size: 20.0f];
    }
    
    return cell;
}

-(void)cancelButtonHit{
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


@end

File: Fitness Tracker/StartWorkout.swift
//
//  StartWorkout.swift
//  Fitness Tracker
//
//  Created by Maruf Nebil Ogunc on 17.08.2018.
//  Copyright © 2018 Maruf Nebil Ogunc. All rights reserved.
//

import UIKit
import AudioToolbox

class StartWorkout: UIViewController, UITableViewDelegate, UITableViewDataSource{
    
    var separatedExercisesArray: NSMutableArray = []
    var countdownTimer: Timer!
    var startPauseWorkoutButton: UIButton!
    var isWorkoutActive = false
    var workoutNo = 0
    var exercisesTableView: UITableView!
    var workoutNameWithoutExtension: String!
    var workoutDict: NSDictionary!
    let jsonManager = JSONManager()

    override func viewDidLoad() {
        super.viewDidLoad()
        self.view.backgroundColor = UIColor.white

        //Getting size of the device
        let screenRect = UIScreen.main.bounds

        //getting workouts list from json
        let workoutsArray = jsonManager.readJSONbyName(name: "workouts")
        
        //removing .json extension
        workoutDict = workoutsArray.object(at: self.workoutNo) as? NSDictionary
        workoutNameWithoutExtension = workoutDict["name"] as? String

        //Customizing navigation bar
        self.navigationController?.navigationBar.titleTextAttributes = [NSAttributedString.Key.foregroundColor: UIColor.black, NSAttributedString.Key.font: UIFont(name: "Metropolis-Bold", size: 20)!]
        self.navigationItem.title = workoutNameWithoutExtension
        
        //Creating cancel button which will close the modal
        let cancelButton = UIBarButtonItem.init(barButtonSystemItem: UIBarButtonItem.SystemItem.cancel, target: self, action: #selector(StartWorkout.cancelButtonHit))
        self.navigationItem.rightBarButtonItem = cancelButton

        //Creating table view which will show exercises
        exercisesTableView = UITableView.init(frame: CGRect(x: 0, y: 0, width: screenRect.size.width, height: screenRect.size.height - 80))
        exercisesTableView.backgroundColor = UIColor.white
        exercisesTableView.delegate = self
        exercisesTableView.dataSource = self
        self.view.addSubview(exercisesTableView)

        //This button will start and pause the workout
        startPauseWorkoutButton = UIButton(type: UIButton.ButtonType.custom)
        startPauseWorkoutButton.setTitle("START", for: UIControl.State.normal)
        startPauseWorkoutButton.setTitleColor(UIColor.white, for: UIControl.State.normal)
        startPauseWorkoutButton.titleLabel?.font = UIFont(name: "Metropolis-Medium", size: 18.0)
        startPauseWorkoutButton.titleLabel?.textAlignment = NSTextAlignment.center
        startPauseWorkoutButton.backgroundColor = UIColor.init(red: 0, green: 179.0/255.0, blue: 85.0/255.0, alpha: 1)
        startPauseWorkoutButton.frame = CGRect(x: 0, y: screenRect.size.height - 80, width: screenRect.size.width, height: 80)
        startPauseWorkoutButton.addTarget(self, action: #selector(StartWorkout.startPauseWorkout), for: UIControl.Event.touchUpInside)
        self.view.addSubview(startPauseWorkoutButton)

        //Get the exercises array from json and separate them
        self.seperateExercisesBySetsAndRests(rawExercisesArray: jsonManager.readJSONbyName(name: workoutDict["name"] as! String))

    }

    @objc func startPauseWorkout() -> () {
        
        if (isWorkoutActive) {
            isWorkoutActive = false;
            startPauseWorkoutButton.backgroundColor = UIColor(red: 0, green: 179.0/255.0, blue: 85.0/255.0, alpha: 1)
            startPauseWorkoutButton.setTitle("START", for: UIControl.State.normal)
            countdownTimer.invalidate()
            
            //changing color of the cell
            let indexOfObject = self.getIndexOfFirstUncompletedExercise()
            let separatedExerciseDict = separatedExercisesArray.object(at: indexOfObject) as! NSDictionary
            let mutableDict: NSMutableDictionary = NSMutableDictionary(dictionary: separatedExerciseDict)
            mutableDict["status"] = "2"
            separatedExercisesArray.replaceObject(at: indexOfObject, with: mutableDict)
            exercisesTableView.reloadData()
            
        }else{
            isWorkoutActive = true;
            startPauseWorkoutButton.backgroundColor = UIColor(red: 229.0/255.0, green: 93.0/255.0, blue: 41.0/255.0, alpha: 1)
            startPauseWorkoutButton.setTitle("PAUSE", for: UIControl.State.normal)
            countdownTimer = Timer.scheduledTimer(timeInterval: 1, target: self, selector: #selector(StartWorkout.updateTimer), userInfo: nil, repeats: true)
        }

    }
    
    func getIndexOfFirstUncompletedExercise() -> Int {
        
        for i in 0..<separatedExercisesArray.count{
            let separatedExerciseDict = separatedExercisesArray.object(at: i) as! NSDictionary
            let durationAsString:String = String(format: "%@", separatedExerciseDict["duration"] as! CVarArg)
            let duration = Int(durationAsString)!

            if(duration>0){
                return i;
            }
        }
        
        return -1
    }
    
    @objc func updateTimer() -> () {
        
        let indexOfObject = self.getIndexOfFirstUncompletedExercise()
        
        if (indexOfObject == -1) {
            countdownTimer.invalidate()
            
            //save the workout info to the user defaults
            jsonManager.saveCompletedWorkout(dict: workoutDict)
            
            let alertController = UIAlertController(title: ("You have completed the " + workoutNameWithoutExtension), message: "You can check your completed workouts on History page", preferredStyle: UIAlertController.Style.alert)
            let cancelAction = UIAlertAction(title: "OK", style: UIAlertAction.Style.cancel, handler: { (action) in
                self.dismiss(animated: true, completion: nil)
            })
            alertController.addAction(cancelAction)
            self.present(alertController, animated: true, completion: nil)
            
            return
        }
        
        //Reducing duration of the exercise
        let separatedExerciseDict = separatedExercisesArray.object(at: indexOfObject) as! NSDictionary
        let durationAsString:String = String(format: "%@", separatedExerciseDict["duration"] as! CVarArg)
        var duration = Int(durationAsString)!
        duration -= 1
        
        if(duration>0){
            
            let separatedExerciseDict = separatedExercisesArray.object(at: indexOfObject) as! NSDictionary
            let mutableDict: NSMutableDictionary = NSMutableDictionary(dictionary: separatedExerciseDict)
            mutableDict["duration"] = String(format: "%d", duration)
            mutableDict["status"] = "1"

            if(duration < 6 && mutableDict["name"] as! String == "Rest"){
                mutableDict["name"] = "Get Ready"
                AudioServicesPlaySystemSoundWithCompletion(kSystemSoundID_Vibrate, {})
            }

            separatedExercisesArray.replaceObject(at: indexOfObject, with: mutableDict)
            exercisesTableView.reloadData()

        }else{
            
            AudioServicesPlaySystemSoundWithCompletion(kSystemSoundID_Vibrate, {})
            
            separatedExercisesArray.removeObject(at: indexOfObject)
            exercisesTableView.reloadData()

        }
    }
    
    func seperateExercisesBySetsAndRests(rawExercisesArray: NSMutableArray) -> () {
        
        for i in 0..<rawExercisesArray.count {
            let exerciseDict = rawExercisesArray.object(at: i) as! NSDictionary

            if(exerciseDict["isCardio"] as! Bool){
                let cardioSeconds = Int(exerciseDict["cardio_minutes"] as! String)! * 60
                let separatedExercise = ["name" : exerciseDict["name"], "duration" : String(format: "%d", cardioSeconds), "status" : "0"]
                separatedExercisesArray.add(separatedExercise)
            }else{
                let sets = Int(exerciseDict["sets"] as! String)!
                
                for j in 0..<sets {
                    let separatedExercise = ["name" : exerciseDict["name"], "duration" : exerciseDict["duration"], "reps" : exerciseDict["reps"], "status" : "0"]
                    separatedExercisesArray.add(separatedExercise)
                    
                    if (j != sets-1){
                        let separatedExercise = ["name" : "Rest", "duration" : workoutDict["rest"], "status" : "0"]
                        separatedExercisesArray.add(separatedExercise)
                    }
                }
            }
            
            //rest between exercises
            if (i != rawExercisesArray.count - 1) {//don't add rest time for last exercise
                
                let separatedExercise = ["name" : "Rest", "duration" : workoutDict["rest"], "status" : "0"]
                separatedExercisesArray.add(separatedExercise)

            }

        }
    }
    
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return separatedExercisesArray.count
    }

    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 70;
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let cellIdentifier = "separatedCell"
        
        var cell: SeparatedExerciseCell? = tableView.dequeueReusableCell(withIdentifier: cellIdentifier) as? SeparatedExerciseCell
        if(cell == nil){
            cell = SeparatedExerciseCell.init(style: UITableViewCell.CellStyle.default, reuseIdentifier: cellIdentifier)
        }
        
        cell?.selectionStyle = UITableViewCell.SelectionStyle.none

        let separatedExerciseDict = separatedExercisesArray.object(at: indexPath.row) as! NSDictionary
        let durationAsString:String = String(format: "%@", separatedExerciseDict["duration"] as! CVarArg)
        let duration = Int(durationAsString)!
        let minutes = duration/60
        let seconds = duration%60
        cell?.countdownLabel.text = String(format: "%d:%02d", minutes, seconds)
        cell?.exerciseNameLabel.text = separatedExerciseDict["name"] as? String
        
        if let reps = separatedExerciseDict["reps"]{
            cell?.repsLabel.isHidden = false
            cell?.repsStaticLabel.isHidden = false
            cell?.repsLabel.text = reps as? String
        }else{
            cell?.repsLabel.isHidden = true
            cell?.repsStaticLabel.isHidden = true
        }

        if (Int(separatedExerciseDict["status"] as! String)! == 0) {//default status
            cell?.backgroundColor = UIColor.white
            cell?.countdownLabel.textColor = UIColor.black
            cell?.exerciseNameLabel.textColor = UIColor.black
            cell?.exerciseNameLabel.font = UIFont(name: "Metropolis-Medium", size: 22.0)
            cell?.repsLabel.textColor = UIColor.black
            cell?.repsStaticLabel.textColor = UIColor.black
        }else if (Int(separatedExerciseDict["status"] as! String)! == 1){//active
            cell?.backgroundColor = UIColor(red: 0, green: 179.0/255.0, blue: 85.0/255.0, alpha: 1)
            cell?.countdownLabel.textColor = UIColor.white
            cell?.exerciseNameLabel.textColor = UIColor.white
            cell?.exerciseNameLabel.font = UIFont(name: "Metropolis-Bold", size: 22.0)
            cell?.repsLabel.textColor = UIColor.white
            cell?.repsStaticLabel.textColor = UIColor.white
        }else if (Int(separatedExerciseDict["status"] as! String)! == 2){//paused
            cell?.backgroundColor = UIColor(red: 229.0/255.0, green: 93.0/255.0, blue: 41.0/255.0, alpha: 1)
            cell?.countdownLabel.textColor = UIColor.white
            cell?.exerciseNameLabel.textColor = UIColor.white
            cell?.exerciseNameLabel.font = UIFont(name: "Metropolis-Bold", size: 22.0)
            cell?.repsLabel.textColor = UIColor.white
            cell?.repsStaticLabel.textColor = UIColor.white
        }

        return cell!

    }
    
    @objc func cancelButtonHit() -> () {
        if (isWorkoutActive) {
            countdownTimer.invalidate()
        }
        self.dismiss(animated: true, completion: nil)
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    


}
